_posts/2023-11-16-quarkus-meets-langchain4j.adoc

```diff
@@ -0,0 +1,265 @@
+
+== Overview
+
+First, let's have a look at the big picture. When integrating an LLM into a Quarkus application, you need to describe what you want the AI to do. Unlike traditional code, you are going to explain the behavior of the AI using natural language. Of course, there are a few techniques to tame the AI, but we will explore that later.
+
+Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus Langchain4j extension provides two mechanisms to extend AI knowledge:
+
+- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
+- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.
+
+The following diagram illustrates the interactions between the LLM, the tools, and the document stores:
+
+// TODO: Add diagram here
+
+== Show me some code!
+
+Alright, enough "bla bla," let's see some code! We are going to use Open AI GPT-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).
+
+First, we need the `quarkus-langchain4j-openai` extension:
+
+[source, xml]
+----
+<dependency>
+    <groupId>io.quarkiverse.langchain4j</groupId>
+    <artifactId>quarkus-langchain4j-openai</artifactId>
+    <version>0.1.0</version> <!-- Update to use the latest version -->
+</dependency>
+----
+
+Once we have the extension, it's time to tell the LLM what we want to do. The Quarkus LangChain4J extension provides a declarative way to describe LLM interactions. The idea is the same as the Quarkus REST client. We model the interaction using an interface annotated with `@RegisterAiService`:
+
+[source, java]
+----
+@RegisterAiService
+public interface TriageService {
+    // methods.
+}
+----
+
+The rest of the application would be able to use the LLM by injecting the `TriageService` interface and calling the methods.
+
+Speaking about methods, that's where the magic happens. You will describe what you want the LLM to do using natural language. First, you start with `@SystemMessage` to define the role and scope. Then, you can use `@UserMessage` to describe the task.
+
+[source, java]
+----
+@SystemMessage("""
```

Comment:
I think I would keep the `class` part. It doesn't take much space and makes things a bit clearer. But YMMV.